torch/_inductor/select_algorithm.py

```diff
@@ -1122,218 +1122,219 @@ def __call__(
-            from .autotune_process import tuning_pool
+        initial_stdout = sys.stdout
+        initial_stderr = sys.stderr
 
-            # do the optional warmup
-            tuning_pool.initialize()
+        def precompile_with_captured_stdout(choice):
+            with restore_stdout_stderr(initial_stdout, initial_stderr):
+                return choice.precompile()
 
-        def do_autotuning(precompile_fn):
-            precompile_start_ts = time.time()
-            precompile_fn()
-            precompile_elapse = time.time() - precompile_start_ts
+        executor = ThreadPoolExecutor(max_workers=num_workers)
 
-            autotune_start_ts = time.time()
-            timings = self.lookup(
-                choices,
-                name,
-                inputs_key,
-                autotune,
-            )
-            autotune_elapse = time.time() - autotune_start_ts
+        futures = {}
+        for c in choices:
+            if hasattr(c, "precompile"):
+                future = executor.submit(precompile_with_captured_stdout, c)
+                futures[future] = c
 
-            if timings and all(
-                not math.isfinite(timing) for timing in timings.values()
-            ):
-                raise NoValidChoicesError
-
-            if make_benchmark_fn.cache_info().currsize:
-                counters["inductor"]["select_algorithm_autotune"] += 1
-
-            if (
-                make_benchmark_fn.cache_info().currsize
-                or log.getEffectiveLevel() == logging.DEBUG
-                or config.trace.log_autotuning_results
+        @functools.lru_cache(None)
+        @restore_stdout_stderr(initial_stdout, initial_stderr)
+        def wait_on_precompile():
+            counters["inductor"]["select_algorithm_precompile"] += 1
+            for future in as_completed(
+                futures,
+                timeout=timeout,
             ):
-                self.log_results(
-                    name, input_nodes, timings, autotune_elapse, precompile_elapse
-                )
-
-            return timings
-
-        precompile_fn = precompile(choices)
-
-        if return_multi_template and (config.max_autotune or config.max_autotune_gemm):
-
-            def get_timings():
-                timings = do_autotuning(precompile_fn)
-                min_extern_choice = float("inf")
-                for choice, timing in timings.items():
-                    if isinstance(choice, ExternKernelCaller):
-                        min_extern_choice = min(min_extern_choice, timing)
-
-                timings = {
-                    choice: time
-                    for choice, time in timings.items()
-                    if (
-                        time <= min_extern_choice
-                        or not isinstance(choice, ExternKernelCaller)
+                if e := future.exception():
+                    log.error(
+                        "Exception %s for benchmark choice %s", e, futures[future]
```

Comment:
Move the precompile logic out of __call__, I think it is a bit easier to follow the control logic of __call__ now